PAYARA-3924: Documentation on Payara BOM (#663)
payara/Payara-Server-Documentation: documentation/ecosystem/maven-bom.adoc

= Bill Of Material Artifact

_Since 5.194_

Payara Platform consists of many artifacts and it depends on lots of APIs as well as their implementations.
Having matching versions of these dependencies prevents many classloading issues.

> Payara's Bill Of Material (BOM) artifact collects all of these version for following types of artifacts:

Reviewer: ```suggestion
Payara's Bill Of Material (BOM) artifact collects all of these versions for the following types of artifacts:
```

[source,type=xml]
```
<dependencyManagement>
    <dependencies>
        <dependency>
            <groupId>fish.payara.api</groupId>
            <artifactId>payara-bom</artifactId>
            <version>5.194</version>
            <type>pom</type>
            <scope>import</scope>
        </dependency>
    </dependencies>
</dependencyManagement>

<!-- Some of the referenced APIs in POM are Payara's patched versions.
      These are published in following repo: -->
<repositories>
    <repository>
        <id>payara-patched-externals</id>
        <name>Payara Patched Externals</name>
        <url>https://raw.github.com/payara/Payara_PatchedProjects/master</url>
        <releases>
            <enabled>true</enabled>
        </releases>
        <snapshots>
            <enabled>false</enabled>
        </snapshots>
    </repository>

</repositories>
```

An example application utilizing the BOM can be seen in repository https://github.com/payara/Payara-Examples[Payara Examples] at path https://github.com/payara/Payara-Examples/tree/master/ecosystem/payara-bom/bom-import[ecosystem/payara-bom/bom-import/].

== Contents of the BOM

=== Payara Distribution

> Distribution artifacts do not define scope, and are therefore imported in scope `compile`..

Reviewer: ```suggestion
Distribution artifacts do not define scope and are therefore imported in scope `compile`..
```
Second part is an dependent clause (cannot stand on its own) that should not be separated by a comma.

| EJB HTTP Client
|
<dependency>
  <groupId>fish.payara.extras</groupId>
  <artifactId>ejb-http-client</artifactId>
</dependency>
|

| Appclient
|
<dependency>
  <groupId>fish.payara.server.appclient</groupId>
  <artifactId>payara-client</artifactId>
</dependency>
|

| Payara API
|
<dependency>
  <groupId>fish.payara.api</groupId>
  <artifactId>payara-api</artifactId>
</dependency>
|

|===

=== APIs

APIs are declared in scope provided, as they define classes present within the server runtime.

[%header, cols="1,3l,2"]
|===
|Artifact | Example usage | Note

| Jakarta EE
|
<dependency>
  <groupId>jakarta.platform</groupId>
  <artifactId>jakarta.jakartaee-api</artifactId>
</dependency>
|Also available `jakarta.jakartaee-web-api`, and all individual API artifacts.

| Eclipse Microprofile
|
<dependency>
  <groupId>org.eclipse.microprofile</groupId>
  <artifactId>microprofile</artifactId>
  <type>pom</type>
</dependency>
| Individual APIs are also declared.

| OSGi
|
<dependency>
    <groupId>org.osgi</groupId>
    <artifactId>org.osgi.core</artifactId>
</dependency>
| Also available are `org.osgi.compendium` and `org.osgi.enterprise`

|===

=== Test tooling

> Arquillian is recomended way for integration tests with the server.

Reviewer: ```suggestion
Arquillian is the recommended way for integration tests with the server.
```
